include/vara/Feature/Constraint.h

```diff
@@ -37,161 +45,216 @@ class NumericConstraint {
 };
 
 class DivisionConstraint : public BinaryConstraint, public NumericConstraint {
 public:
-  DivisionConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  DivisionConstraint(std::unique_ptr<Constraint> LeftOperand,
+                     std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
 class LessConstraint : public BinaryConstraint, public NumericConstraint {
 public:
-  LessConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  LessConstraint(std::unique_ptr<Constraint> LeftOperand,
+                 std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
 class GreaterConstraint : public BinaryConstraint, public NumericConstraint {
 public:
-  GreaterConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  GreaterConstraint(std::unique_ptr<Constraint> LeftOperand,
+                    std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
 class LessEquelsConstraint : public BinaryConstraint, public NumericConstraint {
 public:
-  LessEquelsConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  LessEquelsConstraint(std::unique_ptr<Constraint> LeftOperand,
+                       std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
-class GreaterEquelsConstraint : public BinaryConstraint, public NumericConstraint {
+class GreaterEquelsConstraint : public BinaryConstraint,
+                                public NumericConstraint {
 public:
-  GreaterEquelsConstraint(Constraint *LeftOperand, Constraint *RightOperand)
-      : BinaryConstraint(LeftOperand, RightOperand) {}
+  GreaterEquelsConstraint(std::unique_ptr<Constraint> LeftOperand,
+                          std::unique_ptr<Constraint> RightOperand)
+      : BinaryConstraint(std::move(LeftOperand), std::move(RightOperand)) {}
 
   [[nodiscard]] std::string toString() const override;
 };
 
+class Feature;
+class FeatureModelBuilder;
+
 class PrimaryConstraint : public Constraint {
 public:
-  PrimaryConstraint(Feature *Feature)
-      : Feature(Feature), Constraint(ConstraintKind::CK_PRIMARY) {}
+  PrimaryConstraint(std::variant<Feature *, std::string> FV)
+      : FV(std::move(FV)), Constraint(ConstraintKind::CK_PRIMARY) {}
+
+  std::variant<Feature *, std::string> getFeature() { return FV; }
```

Comment:
This interface seems a bit off. One requests a Feature but also could get a string (currently I'm not sure what the meaning of the string is). Depending on the string meaning, it could be better to check the type in the variant here and return an optional feature (or nullptr) if there was a string in the variant and the `Feature *` otherwise. We could then offer the same interface for the "meaning of string".

This would not tie the user into our variant usage and require all using code to make variant dispatches.

Reply:
Indeed, it seems reasonable to wrap the optional feature within the getter.